Add a developing with spectral-cube docs page (#798)
radio-astro-tools/spectral-cube: docs/developing_with_spectralcube.rst

.. _doc_developersnotes:

Notes for development using spectral-cube
=========================================

> .. currentmodule:: spectral_cube

Reviewer: whooaaaa didn't know you could do this.

Author: Me either! I copy-pasted from docs Chris wrote.

spectral-cube is flexible and can used within other packages for
development beyond the core package's capabilities. Two significant strengths
are the use of memory-mapping and the integration with `dask <https://dask.org/>`_
(:ref:`doc_dask`) to efficiently handle larger than memory data.

> This page provides suggestions for developing using spectral-cube in other

Reviewer: ```suggestion
This page provides suggestions for software development using spectral-cube in other
```

packages.

The following two sections give information on standard usage of  :class:`SpectralCube`.
The third discusses usage with dask integration in :class:`DaskSpectralCube`.

Handling large data cubes
-------------------------

spectral-cube is specifically designed for handling larger-than-memory data
and minimizes creating copies of the data. :class:`SpectralCube` uses memory-mapping
and provides options for executing operations with only subsets of the data
(for example, the `how` keyword in :meth:`SpectralCube.moment`).

Masking operations can be performed "lazily", where the computation is completed
only when a view of the underlying boolean mask array is returned.
See :ref:`doc_masking` for details on these implementations.

Further strategies for handling large data is given in :ref:`doc_handling_large_datasets`.

Parallelizing operations
------------------------

Several operations implemented in :class:`SpectralCube` can be parallelized
using the `joblib <https://joblib.readthedocs.io/en/latest/>`_ package. Builtin methods
in :class:`SpectralCube` with the `parallel` keyword will enable using joblib.

> New methods can take advantage of these features by using creating custom functions

Reviewer: ```suggestion
New methods can take advantage of these features by creating custom functions
```

to pass to :meth:`SpectralCube.apply_function_parallel_spatial` and

> :meth:`SpectralCube.apply_function_parallel_spectral`. These methods expect
> functions with a data and mask array input, with optional `**kwargs` that can be
> passed and expect an output array of the same shape as the input.

Reviewer: ```suggestion
:meth:`SpectralCube.apply_function_parallel_spectral`. These methods accept
functions that take a data and mask array input, with optional `**kwargs`, 
and that return an output array of the same shape as the input.
```
